pkg/service/replay/utils.go

```diff
@@ -69,6 +78,453 @@ func (t *requestMockUtil) SimulateRequest(ctx context.Context, _ uint64, tc *mod
+	case models.Python:
+		coverageData, err = CalPythonCoverage(ctx)
+	case models.Node:
+		coverageData, err = CalTypescriptCoverage()
+	case models.Java:
+		coverageData, err = CalJavaCoverage(logger)
+	}
+	if err != nil {
+		utils.LogError(logger, err, "failed to calculate coverage for the test run")
+	} else {
+		err = reportDB.InsertCoverageReport(ctx, testRunID, &coverageData)
+		if err != nil {
+			utils.LogError(logger, err, "failed to update report with the coverage data")
+		}
+	}
+}
+
+func mergeAndGenerateJacocoReport(ctx context.Context, logger *zap.Logger) error {
+	jacocoPath := filepath.Join(os.TempDir(), "jacoco")
+	jacocoCliPath := filepath.Join(jacocoPath, "jacococli.jar")
+	err := mergeJacocoCoverageFiles(ctx, jacocoCliPath)
+	if err == nil {
+		err = generateJacocoReport(ctx, jacocoCliPath)
+		if err != nil {
+			logger.Debug("failed to generate jacoco report", zap.Error(err))
+		}
+	} else {
+		logger.Debug("failed to merge jacoco coverage data", zap.Error(err))
+	}
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func mergeJacocoCoverageFiles(ctx context.Context, jacocoCliPath string) error {
+	// Find all .exec files starting with "test-set" in the target directory
+	sourceFiles, err := filepath.Glob("target/test-set*.exec")
+	if err != nil {
+		return fmt.Errorf("error finding coverage files: %w", err)
+	}
+	if len(sourceFiles) == 0 {
+		return errors.New("no coverage files found")
+	}
+
+	// Construct the command arguments
+	args := []string{
+		"java",
+		"-jar",
+		jacocoCliPath,
+		"merge",
+	}
+
+	args = append(args, sourceFiles...)
+
+	// Specify the output file
+	args = append(args, "--destfile", "target/keploy-e2e.exec")
+
+	cmd := exec.CommandContext(ctx, args[0], args[1:]...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed to merge coverage files: %w", err)
+	}
+
+	return nil
+}
+
+func generateJacocoReport(ctx context.Context, jacocoCliPath string) error {
+	reportDir := "target/site/keployE2E"
+
+	// Ensure the report directory exists
+	if err := os.MkdirAll(reportDir, 0777); err != nil {
+		return fmt.Errorf("failed to create report directory: %w", err)
+	}
+
+	command := []string{
+		"java",
+		"-jar",
+		jacocoCliPath,
+		"report",
+		"target/keploy-e2e.exec",
+		"--classfiles",
+		"target/classes",
+		"--csv",
+		reportDir + "/e2e.csv",
+		"--html",
+		reportDir,
+	}
+
+	cmd := exec.CommandContext(ctx, command[0], command[1:]...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed to generate report: %w", err)
+	}
+
+	return nil
+}
+
+func CalGoCoverage() (models.TestCoverage, error) {
+	testCov := models.TestCoverage{
+		FileCov:  make(map[string]string),
+		TotalCov: "",
+	}
+
+	coveragePerFileTmp := make(map[string][]int) // filename -> [noOfLines, coveredLines]
+	covdata, err := os.ReadFile(os.Getenv("GOCOVERDIR") + "/total-coverage.txt")
```

Comment:
lets set this if user doesn't set it.

Reply:
https://github.com/keploy/keploy/blob/1806ab70b092d3a3a46cf4919c25f11f8973f3b8/cli/provider/cmd.go#L533

It is done here